cpp/matrix.hpp

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+#include <vector>
+#include <string>
+#include <algorithm>
+
+/**
+ * @brief 行列
+ * @tparam T 型 行列(グリッド)の要素となるintやchar
+ */
+template<class T> struct Matrix {
+    std::vector<std::vector<T>> v;
+    int n, m;
+
+    /**
+     * @brief コンストラクタ
+     * @param v 行列(グリッド)の元となる vector<string> や vector<vector<T>>
+     * @return Matrix
+     */
+    template <typename Iterable>
+    constexpr Matrix(const std::vector<Iterable>& v_) noexcept : n(v_.size()), m(v_.size() == 0 ? 0 : v_[0].size()) {
+        v.resize(n);
+        for(int i = 0; i < n; i++) {
+            v[i].assign(v_[i].begin(), v_[i].end());
+        }
+    }
+
+    /**
+     * @brief コンストラクタ
+     * @param _n 行列(グリッド)の行数
+     * @param _m 行列(グリッド)の列数
+     * @param _val 行列(グリッド)の要素の初期値
+     * @return Matrix
+     */
+    constexpr Matrix(int _n, int _m, T _val) : v(n, std::vector<T>(m, _val)) {}
+    
+    constexpr auto begin() noexcept {return v.begin();}
+    constexpr auto end() noexcept {return v.end();}
+
+    /**
+     * @brief 行列(グリッド)の行数
+     * @return size_t
+     */
+    constexpr size_t size() const {return v.size();}
+    std::vector<T>& operator [] (int i) {return v[i];}
```

Comment:
```suggestion
    std::vector<T>& operator [] (int i) {return v[i];}
    const std::vector<T>& operator [] (int i) const {return v[i];}
```
`const std::vector<T>&` を取る関数に突っ込めないのでconst参照を返す方もいります